### PR TITLE
[CGData][lld-macho] Merge CG Data by LLD

### PR DESCRIPTION
**Pull request by kyulee-com**
LLD now processes raw CG data for stable functions, similar to how it handles raw CG data for the outliner's hash tree. This data is encoded in the custom section (`__llvm_merge`) within object files. LLD merges this information into the indexed CG data file specified by the `-codegen-data-generate-path={path}` option. For the linker that does not support this feature, we could use `llvm-cgdata` tool -- https://github.com/llvm/llvm-project/blob/main/llvm/docs/CommandGuide/llvm-cgdata.rst.

Depends on #115750.
This is a patch for https://discourse.llvm.org/t/rfc-global-function-merging/82608.

### CODE REVIEW

**Comment by kyulee-com** [comment]
cc. @nocchijiang

File: lld/MachO/Driver.cpp [hunk 1]
```diff
@@ -1326,7 +1326,8 @@ static void codegenDataGenerate() {
   TimeTraceScope timeScope("Generating codegen data");
 
   OutlinedHashTreeRecord globalOutlineRecord;
-  for (ConcatInputSection *isec : inputSections)
+  StableFunctionMapRecord globalMergeRecord;
+  for (ConcatInputSection *isec : inputSections) {
     if (isec->getSegName() == segment_names::data &&
```
Comment (alx32) [inline lld/MachO/Driver.cpp:1331]: Can we merge the `if(isec->getSegName() == segment_names::data)` check here and below on 1341 ?

File: lld/MachO/Driver.cpp [hunk 2]
```diff
@@ -1337,10 +1338,25 @@ static void codegenDataGenerate() {
       // Merge it to the global hash tree.
       globalOutlineRecord.merge(localOutlineRecord);
     }
+    if (isec->getSegName() == segment_names::data &&
+        isec->getName() == section_names::functionmap) {
+      // Read stable functions from each section.
+      StableFunctionMapRecord localMergeRecord;
+      auto *data = isec->data.data();
```
Comment (alx32) [inline lld/MachO/Driver.cpp:1345]: any reason `data` here is a separate variable ?

File: lld/MachO/InputSection.h [hunk 3]
```diff
@@ -339,6 +339,7 @@ constexpr const char const_[] = "__const";
 constexpr const char lazySymbolPtr[] = "__la_symbol_ptr";
 constexpr const char lazyBinding[] = "__lazy_binding";
 constexpr const char literals[] = "__literals";
+constexpr const char functionmap[] = "__llvm_merge";
```
Comment (alx32) [inline lld/MachO/InputSection.h:342]: I think `functionMap` would be the name here if we were to match existing pattern.

File: lld/MachO/Driver.cpp (hunk 2 above)
Comment (kyulee-com) [reply lld/MachO/Driver.cpp:1345]: Yes because of `void deserialize(const unsigned char *&Ptr)` which modifies/advances the pointer. I added a comment above.

File: lld/test/MachO/cgdata-generate-merge.s [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# UNSUPPORTED: system-windows
+# REQUIRES: aarch64
+
+# RUN: rm -rf %t; split-file %s %t
```
Comment (MaskRay) [inline lld/test/MachO/cgdata-generate-merge.s:4]: drive-by: you can add `cd %t` to remove `%t/` below

Comment (ellishg) [reply lld/test/MachO/cgdata-generate-merge.s:4]: I actually like keeping the `%t/` in the commands because it allows me to copy and run the commands directly for debugging without needing to first `cd %t`.

**Comment by alx32** [comment]
LGTM !

**Review by alx32 — APPROVED** [review]

File: lld/test/MachO/cgdata-generate-merge.s (hunk 4 above)
Comment (kyulee-com) [reply lld/test/MachO/cgdata-generate-merge.s:4]: Yes, I'll leave it as is, since each command is self-contained and complete, making it convenient to reproduce.
